### PR TITLE
An initial VilBERT model for NLVR2

### PR DESCRIPTION
**Pull request by matt-gardner**

### CODE REVIEW

File: allennlp/data/dataset_readers/nlvr2_reader.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+                new_question += question[prev_end:idx]
+                prev_end = idx + length
+            new_question += question[prev_end:]
+            question = new_question
+        tokenized_sentence = self._tokenizer.tokenize(question)
+        tokenized_sentence = [str(tok) for tok in tokenized_sentence]
+        sentence_field = TextField(
+            [Token(self._tokenizer.cls_token)]
+            + [Token(tok) for tok in tokenized_sentence]
+            + [Token(self._tokenizer.sep_token)],
```
Comment (dirkgr) [inline allennlp/data/dataset_readers/nlvr2_reader.py:192]: You should call `self._tokenizer.add_special_tokens()` for this.

Comment (jiasenlu) [reply allennlp/data/dataset_readers/nlvr2_reader.py:192]: For the dataset_readers, do we want to create the reader for each dataset?  (VQA, GQA, refer expression), tasks, or models?

Comment (matt-gardner) [reply allennlp/data/dataset_readers/nlvr2_reader.py:192]: Yes, each dataset would need a dataset reader.  If suitable readers already exist in other places, we can re-use as much code as possible (though we'd definitely have to write at least a little bit of code to get them into allennlp's data pipeline).

**Comment by matt-gardner** [comment]
Given that: (1) we're largely in agreement about the direction to go here, (2) we're now talking about various improvements on top of this code, which can be done independently, and (3) this is going to be merged into a branch, not `master`, I propose that we merge this now, as it is, and continue making PRs to the vision branch.  @dirkgr, I'll convert this from a draft to a real PR, and if you're happy with this plan, feel free to merge when you start working on things.

**Comment by matt-gardner** [comment]
Also, to be clear, I think we should change almost all of the names of all the classes in this PR, and there's a lot of other stuff I want to clean up.  But we can do that in future PRs and unblock people from making orthogonal improvements to what's here.

**Comment by dirkgr** [comment]
@epwalsh, do you know why this isn't running any tests? Is it because we're merging into the `vision` branch, not the `master` branch? Can we change it so tests run on this too?

File: allennlp/models/vilbert.py [hunk 2]
```diff
@@ -0,0 +1,836 @@
...
+        )  # fp16 compatibility
+
+        extended_image_attention_mask = extended_image_attention_mask.to(
+            dtype=next(self.parameters()).dtype
+        )  # fp16 compatibility
+        extended_image_attention_mask = (1.0 - extended_image_attention_mask) * -10000.0
+
+        co_attention_mask = torch.zeros(batch_size, num_images, feature_size, num_tokens).type_as(
+            extended_image_attention_mask
+        )
```
Comment (dirkgr) [inline allennlp/models/vilbert.py:787]:
```suggestion
        co_attention_mask = torch.zeros(batch_size, num_images, feature_size, num_tokens, dtype=extended_image_attention_mask.dtype)
```
This avoids another temporary tensor.

File: allennlp/models/vilbert.py [hunk 3]
```diff
@@ -0,0 +1,836 @@
...
+        sequence_output_v = encoded_layers_v[:, :, :, :, -1]
+
+        pooled_output_t = self.t_pooler(sequence_output_t)
+        pooled_output_v = self.v_pooler(sequence_output_v)
+
+        if self.fusion_method == "sum":
+            pooled_output = self.dropout(pooled_output_t + pooled_output_v)
+        elif self.fusion_method == "mul":
+            pooled_output = self.dropout(pooled_output_t * pooled_output_v)
+
```
Comment (dirkgr) [inline allennlp/models/vilbert.py:812]: This is missing an `else: raise ValueError(...)`.

**Comment by dirkgr** [comment]
Not merging this right away doesn't have to block us. We can merge this `lxmert` branch into our working branches as it is and start coding. We don't have to wait until it's merged into `vision`. GitHub will sort out the diff automatically when it's time to merge. This works mostly right even if we squash-merge.

**Comment by epwalsh** [comment]
@dirkgr if we add this branch to this list [here](https://github.com/allenai/allennlp/blob/master/.github/workflows/pull_request.yml#L6), the workflows should start running

File: allennlp/modules/time_distributed.py [hunk 4]
```diff
@@ -56,8 +56,20 @@ def forward(self, *inputs, pass_through: List[str] = None, **kwargs):
...
-        outputs = reshaped_outputs.contiguous().view(new_size)
+        tuple_output = True
+        if not isinstance(reshaped_outputs, tuple):
+            tuple_output = False
+            reshaped_outputs = (reshaped_outputs,)
+
+        outputs = []
+        for reshaped_output in reshaped_outputs:
+            print(some_input.size())
+            print(reshaped_output.size())
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/modules/time_distributed.py:67]: Oops, thanks for the catch.

File: allennlp/models/vilbert.py [hunk 5]
```diff
@@ -0,0 +1,836 @@
...
+        visual_features: torch.Tensor,
+        box_coordinates: torch.Tensor,
+        image_id: List[List[str]],
+        identifier: List[str],
+        sentence_field: TextFieldTensors,
+        denotation: torch.Tensor = None,
+    ) -> Dict[str, torch.Tensor]:
+
+        batch_size, num_images, num_objects, feature_size = visual_features.size()
+        assert num_objects == 36 and feature_size == 2048
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/models/vilbert.py:738]: An assert that Sanjay had to be sure his data was read correctly, I think.  I removed it.

File: allennlp/models/vilbert.py [hunk 6]
```diff
@@ -0,0 +1,836 @@
...
+        assert num_images == 2
+
+        input_ids = sentence_field["tokens"]["token_ids"]
+        token_type_ids = sentence_field["tokens"]["type_ids"]
+        attention_mask = sentence_field["tokens"]["mask"]
+        # We will always have an equal number of images and boxes, so no masking necessary.
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/models/vilbert.py:744]: It's about padding; there is no padding in these dimensions.  I clarified the comment.

File: allennlp/models/vilbert.py [hunk 7]
```diff
@@ -0,0 +1,836 @@
...
+        image_attention_mask = torch.ones_like(box_coordinates)[:, :, :, 0]
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/models/vilbert.py:745]: Yep, good point.

File: allennlp/models/vilbert.py (hunk 2 above)
Comment (matt-gardner) [reply allennlp/models/vilbert.py:787]: Done.

File: allennlp/models/vilbert.py (hunk 3 above)
Comment (matt-gardner) [reply allennlp/models/vilbert.py:812]: Done.

**Comment by matt-gardner** [comment]
@dirkgr, this is ready for another look.

File: allennlp/models/vilbert.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dirkgr) [reply allennlp/models/vilbert.py]: Oh, I see. #objects and #boxes are always equal to themselves, not to each other.

**Comment by dirkgr** [comment]
GitHub refuses to show me anything but a 1000-line diff with 90% formatting changes. Since it's not going into `master`, and I think I skimmed the actual changes yesterday, I'll merge this.
